tools/distpartitioning/multi_dev_init.py

```diff
@@ -0,0 +1,499 @@
+    trigger alltoallv_cpu for the message exchange.
+
+    Parameters:
+    -----------
+    rank : int
+        rank of the current process
+    world_size : int
+        total no. of participating processes
+    node_data : dictionary
+        nodes data dictionary with keys as column names and values as
+        columns from the nodes csv file
+    """
+    input_list = []
+    send_sizes = []
+    recv_sizes = []
+    start = timer()
+    for i in np.arange(world_size):
+        send_idx = (node_data[constants.OWNER_PROCESS] == i)
+        idx = send_idx.reshape(node_data[constants.GLOBAL_NID].shape[0])
+        filt_data = np.column_stack((node_data[constants.NTYPE_ID][idx == 1], \
+                                node_data[constants.GLOBAL_TYPE_NID][idx == 1], \
+                                node_data[constants.GLOBAL_NID][idx == 1]))
+        if(filt_data.shape[0] <= 0): 
+            input_list.append(torch.empty((0,), dtype=torch.int64))
+            send_sizes.append(torch.empty((0,), dtype=torch.int64))
+        else:
+            input_list.append(torch.from_numpy(filt_data))
+            send_sizes.append(torch.tensor(filt_data.shape, dtype=torch.int64))
+        recv_sizes.append(torch.zeros((2,), dtype=torch.int64))
+    end = timer()
+    print('[Rank: ', rank, '] Preparing node_data to send out: ', timedelta(seconds=end - start))
+
+    #exchange sizes first followed by data. 
+    dist.barrier()
+    start = timer()
+    alltoall_cpu(rank, world_size, recv_sizes, send_sizes)
+
+    output_list = []
+    for s in recv_sizes: 
+        output_list.append(torch.zeros(s.tolist(), dtype=torch.int64))
+    
+    dist.barrier()
+    alltoallv_cpu(rank, world_size, output_list, input_list)
+    end = timer()
+    print('[Rank: ', rank, '] Time to exchange node data : ', timedelta(seconds=end - start))
+
+    #stitch together the received data to form a consolidated data-structure
+    rcvd_node_data = torch.cat(output_list).numpy()
+    print('[Rank: ', rank, '] Received node data shape ', rcvd_node_data.shape)
+
+    #Replace the node_data values with the received node data and the OWNER_PROCESS key-value
+    #pair is removed after the data communication
+    node_data[constants.NTYPE_ID] = rcvd_node_data[:,0]
+    node_data[constants.GLOBAL_TYPE_NID] = rcvd_node_data[:,1]
+    node_data[constants.GLOBAL_NID] = rcvd_node_data[:,2]
+    node_data.pop(constants.OWNER_PROCESS)
+
+def exchange_edge_data(rank, world_size, edge_data):
+    """
+    Exchange edge_data among processes in the world.
+    Prepare list of sliced data targeting each process and trigger
+    alltoallv_cpu to trigger messaging api
+
+    Parameters:
+    -----------
+    rank : int
+        rank of the process
+    world_size : int
+        total no. of processes
+    edge_data : dictionary
+        edge information, as a dicitonary which stores column names as keys and values
+        as column data. This information is read from the edges.txt file.
```

Comment:
edge_data is used for both input and output. this needs to be documented.
however, this is not a right method. you should return the results instead of modifying the input data.

Reply:
Agree. Will change the code as suggested. This means that all the following functions will change: "exchange_node_data, exchange_edge_data and exchange_node_features". I will make these code changes and followup with the code drop in a couple of hours.

Reply:
yes, all functions should be changed to return new dicts.

Reply:
Agreed. This will be done.